Core/clim-basic/utils.lisp

```diff
@@ -743,3 +728,195 @@ a flag CLOSED is T then beginning and end of the list are consecutive too."
+                    (return-from breaks-rec
+                      (breaks-rec offset margin start end (1+ next-opportunity-index)))))
+             (loop
+                with best-breaks = nil
+                with best-cost = nil
+                for i from next-opportunity-index below (length opportunities)
+                for opportunity = (aref opportunities i)
+                for current-width = (funcall width-fn string start opportunity)
+                do
+                  (cond ((<= current-width current-margin)
+                         (let ((current-cost (expt (- current-margin current-width) 2))
+                               (current-break (if-let ((pos (skip-whitespace string opportunity)))
+                                                pos
+                                                (return-from breaks-rec nil))))
+                           (multiple-value-bind (breaks remaining-cost)
+                               (breaks-rec 0 margin current-break end (1+ i))
+                             (when (or (null best-cost)
+                                       (< (+ current-cost remaining-cost) best-cost))
+                               (setf best-breaks (list* current-break breaks)
+                                     best-cost (+ current-cost remaining-cost))))))
+                        ((null best-cost)
+                         (return-from breaks-rec
+                           (if (> offset 0)
+                               (multiple-value-bind (breaks remaining-cost)
+                                   (breaks-rec 0 margin start end i)
+                                 (return-from breaks-rec
+                                   (values (list* start breaks)
+                                           (+ remaining-cost current-margin))))
+                               (let* ((char-breaks (%line-breaks-1 string width
+                                                                   offset margin
+                                                                   start opportunity))
+                                      (new-start (alexandria:last-elt char-breaks)))
+                                 (multiple-value-bind (breaks remaining-cost)
+                                     (breaks-rec 0 margin new-start end i)
+                                   (values (append char-breaks breaks) remaining-cost))))))
+                        (T #1=(return-from breaks-rec
+                                (values best-breaks best-cost))))
+                finally #1#)))
+    (breaks-rec initial-offset margin start end 0)))
+
+;; Implementing line breaking as defined in Unicode[1] is left as an excercise
+;; for the reader. [1] https://unicode.org/reports/tr14/ -- jd 2019-01-08
+(defun line-breaks (string width &key (break-strategy t) initial-offset margin (start 0) end)
+  "Function takes a string and returns a list of indexes where it should be split.
```

Comment:
Can the string contain #\Newline characters? If so, how are they handled?

Reply:
It can and how they are handled is not a concern of `line-breaks` (it all depends on `break-strategy` - i.e newline may be counted as a "break" character). The utility doesn't put any meaning to the characters (it is a responsibility of the break-strategy). Unicode algorithm should take such things into account (hard and soft breaks).

From the wider perspective (McCLIM-wise): stream-write-string splits the string by newline first so it is always the case, that there are no newlines in a string passed to `line-breaks`

```
(defmethod stream-write-string ((stream standard-extended-output-stream) string
                                &optional (start 0) end)
  (let ((seg-start start)
        (end (or end (length string))))
    (with-cursor-off stream
      (loop for i from start below end do
           (when (char= #\Newline
                        (char string i))
             (seos-write-string stream string seg-start i)
             (seos-write-newline stream)
             (setq seg-start (1+ i))))
      (seos-write-string stream string seg-start end))))
```

All breaks in seos-write-string are in fact soft newline breaks (candidates for rewrapping). I plan to use that quality when I get to the rewrapping functionality.

Reply:
I think the first paragraph of what you explained here is a good candidate for inclusion in the documentation string.

Reply:
"default line breaking by word strategy" is a purposefully vague wording. Break opportunities are one thing, but by default we may want to break with minimum raggedness too (for instance), or we want to switch to some other approach (which is not related to unicode). Also adding into a docstring something what we wish we were doing misses the purpose of the documentation (imo).

Reply:
I meant this first paragraph:

> It can [contain newlines] and how they are handled is not a concern of line-breaks [...]

Mentioning this in the documentation string would make sense in my opinion. I was wondering about that after reading the documentation string.

Reply:
ah, makes sense. force-pushed (the last docstring sentence mentions it now).